core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheManager.java

```diff
@@ -445,18 +445,33 @@ private boolean deletePage(PageId pageId, PageInfo pageInfo) {
     return true;
   }
 
-  @Nullable
-  private ReadableByteChannel getPage(PageId pageId, int offset) {
-    ReadableByteChannel ret;
-    try {
-      ret = mPageStore.get(pageId, offset);
+  private int getPage(PageId pageId, int offset, int bytesToRead, byte[] buffer,
+      int offsetInBuffer) {
+    try (ReadableByteChannel chan = mPageStore.get(pageId, offset)) {
+      // wrap return byte array in a bytebuffer and set the pos/limit for the page read
+      ByteBuffer buf = ByteBuffer.wrap(buffer);
+      buf.position(offsetInBuffer);
+      buf.limit(offsetInBuffer + bytesToRead);
+      // read data from cache
+      while (buf.position() != buf.limit()) {
+        if (chan.read(buf) == -1) {
+          break;
+        }
+      }
+      if (buf.position() != buf.limit()) {
+        // data read from page store is inconsistent from the metastore
+        Metrics.GET_ERRORS_FAILED_READ.inc();
+        throw new IOException(String.format(
+            "Failed to read page {}: supposed to read {} bytes, {} bytes actually read",
+            pageId, bytesToRead, buf.position() - offsetInBuffer));
+      }
     } catch (IOException | PageNotFoundException e) {
       LOG.error("Failed to get existing page {}: {}", pageId, e);
       Metrics.GET_ERRORS.inc();
-      return null;
+      return -1;
```

Comment:
Should we have this logic here or leave it to `NoExceptionCacheManager`?

Reply:
it needs to be here so we can clean up the state https://github.com/Alluxio/alluxio/pull/11525/files#diff-52b3926ca60dbb1ab7f30f23ada07efdR365

Reply:
got it, thanks.